# 위키피디아 데이터에서 문서 찾기

## Import Graphlab

graphlab을 import합니다. 에러가 나는 경우 'install graphlab' 노트북을 열어서 설치해야 합니다.

In [10]:
#import graphlab

## 데이터 화일 업로드하기

먼저 people_wiki.gl.zip 화일을 업로드합니다. Destination Folder에 /library/data 폴더를 지정합니다.

ls 명령을 이용해서 data 폴더 안에 people_wiki.gl.zip 화일이 제대로 업로드되었는지 확인합시다.

In [11]:
#ls ../data

이제 unzip 명령을 이용해서 압축화일을 해제합시다.

In [12]:
#!unzip -o ../data/people_wiki.gl.zip -d ../data

# 위키피디아에서 인물에 대한 텍스트 데이터를 로드합니다.

위키피디아의 인물에 대한 텍스트 데이터를 읽어 들입니다.

SFrame 데이터를 읽어 와서 **people**이라는 변수에 저장합니다.
**'../data/people_wiki.gl/'**이라는 폴더를 지정해야 합니다.

    graphlab.SFrame(폴더위치)

In [13]:
people = graphlab.SFrame('../data/people_wiki.gl/')

head 메소드로 people에 어떤 데이터가 들어 있는지 확인해 봅시다.

In [14]:
people.head()

Columns:
	URI	str
	name	str
	text	str

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+
|              text             |
+-------------------------------+
| digby morrell born 10 octo... |
| alfred j lewy aka sandy le... |
| harpdog brown is a singer ... |
| franz rottensteiner born i... |
| henry krvits born 30 decem... |
| sam henderson born october... |
| aaron lacrate is an americ... |
| trevor ferguson aka john f... |
| grant nelson born 27 april... |
| cathy caruth born 1955 is ... |
+-------------------------------+
[10 rows x 3 columns]

위키피디아 문서로 가는 링크, 인물의 이름, 문서의 텍스트 등이 들어 있습니다.

총 몇명의 데이터인지 확인해 봅시다.

In [15]:
len(people)

59071

59,071명이 맞나요?

# 데이터셋을 살펴보고 어떤 텍스트가 있는지 확인해 봅시다.

## Obama 대통령의 데이터를 확인해 봅시다.

전체 인물 데이터에서 Obama 대통령에 해당하는 데이터를 찾아서 **obama**에 저장합시다. 이름에서 'Barack Obama'를 찾으면 됩니다.

In [16]:
obama = people[people['name'] == 'Barack Obama']

obama의 데이터를 확인해 봅시다.

In [17]:
obama.head()

Columns:
	URI	str
	name	str
	text	str

Rows: 1

Data:
+-------------------------------+--------------+-------------------------------+
|              URI              |     name     |              text             |
+-------------------------------+--------------+-------------------------------+
| <http://dbpedia.org/resour... | Barack Obama | barack hussein obama ii br... |
+-------------------------------+--------------+-------------------------------+
[1 rows x 3 columns]

obama의 **'text'** 컬럼의 데이터를 확인해 봅시다.

In [18]:
obama['text']

dtype: str
Rows: ?
['barack hussein obama ii brk husen bm born august 4 1961 is the 44th and current president of the united states and the first african american to hold the office born in honolulu hawaii obama is a graduate of columbia university and harvard law school where he served as president of the harvard law review he was a community organizer in chicago before earning his law degree he worked as a civil rights attorney and taught constitutional law at the university of chicago law school from 1992 to 2004 he served three terms representing the 13th district in the illinois senate from 1997 to 2004 running unsuccessfully for the united states house of representatives in 2000in 2004 obama received national attention during his campaign to represent illinois in the united states senate with his victory in the march democratic party primary his keynote address at the democratic national convention in july and his election to the senate in november he began his presidential campa

## **대배우님**, George Clooney의 데이터를 살펴 봅시다.

'George Clooney'의 데이터를 찾아서 **clooney**에 저장합시다.

In [19]:
clooney = people[people['name'] == 'George Clooney']

text 컬럼의 데이터를 확인해 봅시다.

In [20]:
clooney['text']

dtype: str
Rows: ?
['george timothy clooney born may 6 1961 is an american actor writer producer director and activist he has received three golden globe awards for his work as an actor and two academy awards one for acting and the other for producingclooney made his acting debut on television in 1978 and later gained wide recognition in his role as dr doug ross on the longrunning medical drama er from 1994 to 1999 for which he received two emmy award nominations while working on er he began attracting a variety of leading roles in films including the superhero film batman robin 1997 and the crime comedy out of sight 1998 in which he first worked with a director who would become a longtime collaborator steven soderbergh in 1999 clooney took the lead role in three kings a wellreceived war satire set during the gulf warin 2001 clooneys fame widened with the release of his biggest commercial success the heist comedy oceans eleven the first of the film trilogy a remake of the 1960 film wit

# Obama 문서의 단어 카운트를 만들어 봅시다.


이제 아래 메소드를 사용해서 Obama 문서의 단어 개수 벡터를 만든 다음, **word_count**라는 컬럼으로 obama에 저장해 봅시다.

    graphlab.text_analytics.count_words(SArray)

주의: 이 메소드는 SFrame이 아니라 SArray를 입력으로 받습니다.

In [24]:
obama['word_count'] = graphlab.text_analytics.count_words(obama['text'])

**'word_count'** 컬럼을 출력해서 내용을 살펴 봅시다.

In [25]:
print obama['word_count']

[{'operations': 1, 'represent': 1, 'office': 2, 'unemployment': 1, 'is': 2, 'doddfrank': 1, 'over': 1, 'unconstitutional': 1, 'domestic': 2, 'named': 1, 'ending': 1, 'ended': 1, 'proposition': 1, 'seats': 1, 'graduate': 1, 'worked': 1, 'before': 1, 'death': 1, '20': 2, 'taxpayer': 1, 'inaugurated': 1, 'obamacare': 1, 'civil': 1, 'mccain': 1, 'to': 14, '4': 1, 'policy': 2, '8': 1, 'has': 4, '2011': 3, '2010': 2, '2013': 1, '2012': 1, 'bin': 1, 'then': 1, 'his': 11, 'march': 1, 'gains': 1, 'cuba': 1, 'californias': 1, '1992': 1, 'new': 1, 'not': 1, 'during': 2, 'years': 1, 'continued': 1, 'presidential': 2, 'husen': 1, 'osama': 1, 'term': 3, 'equality': 1, 'prize': 1, 'lost': 1, 'stimulus': 1, 'january': 3, 'university': 2, 'rights': 1, 'gun': 1, 'republican': 2, 'rodham': 1, 'troop': 1, 'withdrawal': 1, 'involvement': 3, 'response': 3, 'where': 1, 'referred': 1, 'affordable': 1, 'attorney': 1, 'school': 3, 'senate': 3, 'house': 2, 'national': 2, 'creation': 1, 'related': 1, 'hawaii': 1,

관련 문서의 텍스트를 분석하여 등장한 단어를 키(key)로, 단어가 나타난 개수를 값(value)으로 하는 dictionary 형태임을 알 수 있습니다.

## Obama 문서의 단어 카운트를 소팅하기

### 단어 카운트 dictionary를 테이블로 변환하기
word_count 컬럼에 저장되어 있는 dictionary를 stack 메소드를 이용해서 단어를 'word' 컬럼에, 횟수를 'count' 컬럼에 넣습니다. stack 메소드는 주어진 SFrame에 새로운 컬럼들을 추가해서 **각 단어별로** 반환합니다.

    SFrame.stack(컬럼이름, new_column_name=새로운컬럼이름리스트)

주의: **obama 문서의 'word_count' 컬럼만으로 만들어진 SFrame**을 이용하세요.

In [26]:
obama[['word_count']].stack('word_count', new_column_name=['word', 'count'])

Columns:
	word	str
	count	int

Rows: 273

Data:
+------------+-------+
|    word    | count |
+------------+-------+
|    cuba    |   1   |
| relations  |   1   |
|   sought   |   1   |
|   combat   |   1   |
|   ending   |   1   |
| withdrawal |   1   |
|   state    |   1   |
|  islamic   |   1   |
|     by     |   1   |
|   gains    |   1   |
+------------+-------+
[273 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

이 결과를 **obama_word_count_table**에 저장하세요.

In [27]:
obama_word_count_table = obama[['word_count']].stack('word_count', new_column_name=['word', 'count'])

### 맨위에 가장 많이 나오는 단어가 나오도록 단어 카운트 소팅하기

먼저 head 메소드를 이용해서 obama_word_count_table의 내용을 확인합시다.

In [28]:
obama_word_count_table.head()

Columns:
	word	str
	count	int

Rows: 10

Data:
+------------+-------+
|    word    | count |
+------------+-------+
|    cuba    |   1   |
| relations  |   1   |
|   sought   |   1   |
|   combat   |   1   |
|   ending   |   1   |
| withdrawal |   1   |
|   state    |   1   |
|  islamic   |   1   |
|     by     |   1   |
|   gains    |   1   |
+------------+-------+
[10 rows x 2 columns]

이제 'count' 컬럼의 값에 따라 내림차순으로 소팅해 주세요.

    SFrame.sort(컬럼이름, ascending=True또는False)
    

In [29]:
obama_word_count_table.sort('count', ascending=False)

Columns:
	word	str
	count	int

Rows: 273

Data:
+-------+-------+
|  word | count |
+-------+-------+
|  the  |   40  |
|   in  |   30  |
|  and  |   21  |
|   of  |   18  |
|   to  |   14  |
|  his  |   11  |
| obama |   9   |
|  act  |   8   |
|   he  |   7   |
|   a   |   7   |
+-------+-------+
[273 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

많이 등장하는 단어들에는 정보로서의 가치가 없는 단어들 - "the", "in", "and" 등 - 이 다수 포함되어 있습니다.

# Corpus의 TF-IDF 계산하기 

정보로서 가치가 있는 단어들에 더 많은 weight를 주기 위해 TF-IDF 점수를 이용하고자 합니다.

개별 인물이 아닌, 전체 corpus에 대해서 적용해 봅시다.

아래 메소드를 이용해서 people 문서의 텍스트에 대한 단어 개수 벡터를 만든 다음, **word_count**라는 컬럼으로 people에 저장해 봅시다.

    graphlab.text_analytics.count_words(SArray)

주의: 이 메소드는 SFrame이 아니라 SArray를 입력으로 받습니다.

In [30]:
people['word_count'] = graphlab.text_analytics.count_words(people['text'])

head 메소드를 이용해서 내용을 확인합시다.

In [31]:
people.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'selection': 1, 'carltons... |
| alfred j lewy aka sandy le... | {'precise': 1, 'thomas': 1... |
| harpdog brown is a singer ... | {'just': 1, 'issued': 1, '... |
| franz rottensteiner born i... | {'all': 1, 'bauforschung':... |
| henry krvits born 30 decem... | {'they': 1, 'gangstergenka... |
| sam henderson born october... | {'currently': 1, 'less': 1... |
| aaron lacrate is an americ... | {'exclusive': 2, 'producer... |
| trevor ferguson aka john f... | {'taxi': 1, 'salon': 1, 'g... |
| grant nelson born 27 april... | {'houston': 1, 'frankie': ... |
| cathy caruth born 1955 is ... | {'phenomenon': 1, 'deboras... |
+-------------------------------+-------------------------------+
[10 rows x 4 columns]

아래 메소드에 **word_count**를 인자로 넣어서 TF_IDF를 구한 다음, **tdidf**라는 컬럼으로 people에 저장합시다.
    
    graphlab.text_analytics.tf_idf(SArray)
    
주의: 이 메소드는 SFrame이 아니라 SArray를 입력으로 받습니다.

In [32]:
people['tfidf'] = graphlab.text_analytics.tf_idf(people['word_count'])

마찬가지로 head로 내용을 확인합시다.

In [33]:
people.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'selection': 1, 'carltons... |
| alfred j lewy aka sandy le... | {'precise': 1, 'thomas': 1... |
| harpdog brown is a singer ... | {'just': 1, 'issued': 1, '... |
| franz rottensteiner born i... | {'all': 1, 'bauforschung':... |
| henry krvits born 30 decem... | {'they': 1, 'gangstergenka... |
| sam henderson born october... | {'currently': 1, 'less': 1... |
| aaron lacrate is an americ... | {'exclusive': 2, 'producer... |
| trevor ferguson aka john f... | {'taxi': 1, 'salon': 1, 'g... |
| grant nelson born 27 april... | {'houston': 1, 'frankie': ... |
| cathy caruth born 1955 is ... | {'phenomenon': 1, 'deboras... |
+-------------------------------+-------------------------------+
+-------------------------------+
|             tfidf             |
+-------------------------------+
| {'selection': 3.8365785530... |
| {'precise': 6.443200606955... |
| {'just': 2.700729968710864... |
| {'all': 1.6431112434912472... |
| {'they': 1.899340117819389... |
| {'currently': 1.6370889691... |
| {'exclusive': 10.455187230... |
| {'taxi': 6.052021456094502... |
| {'houston': 3.935505942157... |
| {'phenomenon': 5.750053426... |
+-------------------------------+
[10 rows x 5 columns]

## Obama 문서의 TF-IDF 살펴보기

people에서 Obama 문서를 골라내서 **obama**라는 변수에 저장합시다. 'Barack Obama'를 사용하세요.

In [34]:
obama = people[people['name'] == 'Barack Obama']

obama를 출력해 봅시다.

In [35]:
obama.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: 1

Data:
+-------------------------------+--------------+-------------------------------+
|              URI              |     name     |              text             |
+-------------------------------+--------------+-------------------------------+
| <http://dbpedia.org/resour... | Barack Obama | barack hussein obama ii br... |
+-------------------------------+--------------+-------------------------------+
+-------------------------------+-------------------------------+
|           word_count          |             tfidf             |
+-------------------------------+-------------------------------+
| {'operations': 1, 'represe... | {'operations': 3.811771079... |
+-------------------------------+-------------------------------+
[1 rows x 5 columns]

### TF-IDF dictionary를 테이블로 변환하기
'tfidf'컬럼에 저장되어 있는 dictionary를 stack 메소드를 이용해서 단어를 'word' 컬럼에, tf-idf점수를 'tfidf' 컬럼에 넣어 봅시다. stack 메소드는 주어진 SFrame에 새로운 컬럼들을 추가해서 **각 단어별로** 반환합니다.

    SFrame.stack(컬럼이름, new_column_name=새로운컬럼이름리스트)

주의: **obama 문서의 'tfidf' 컬럼만으로 만들어진 SFrame**을 이용하세요.

In [36]:
obama_tfidf = obama[['tfidf']].stack('tfidf', new_column_name=['word', 'tfidf'])

TF-IDF 점수가 높은 단어들이 훨씬 더 정보로서 가치가 높습니다.

# 유명인사들 간의 거리를 직접 계산해 보기

몇몇 유명인사들의 문서들 간의 거리를 직접 계산해 봅시다.

먼저 'Bill Clinton'을 찾아서 **clinton**에 저장합시다.

In [37]:
clinton = people[people['name'] == 'Bill Clinton']

먼저 'David Beckham'을 찾아서 **beckham**에 저장합시다.

In [38]:
beckham = people[people['name'] == 'David Beckham']

## Obama가 Clinton에 가까울까요? 아니면 Beckham에 더 가까을까요?

cosine 거리를 사용하겠습니다. 이는 아래식으로 표현됩니다.

    distance = 1 - cosine_similarity

Obama 대통령의 문서가 유명한 축구선수인 David Beckham의 문서보다 Clinton 대통령의 문서에 가까운지 알아 보겠습니다.
아래 메소드를 사용합니다.

    graphlab.distances.cosine(문서1의tfidf, 문서2의tfidf)
    
주의: 이 메소드는 SArray가 아닌 dictionary 또는 list을 인자로 받습니다.

먼저 obama와 clinton 간의 거리를 확인하세요.

In [39]:
graphlab.distances.cosine(obama[0]['tfidf'], clinton[0]['tfidf'])

0.8339854936884276

그 다음 obama와 beckham 간의 거리를 확인하세요.

In [40]:
graphlab.distances.cosine(obama[0]['tfidf'], beckham[0]['tfidf'])

0.9791305844747478

누가 더 obama에 가깝나요?

# 문서 조회를 위해 최근접 이웃 (nearest neighbor) 모델 만들기

이제 nearest neighbor 모델을 만들어서 문서 조회에 이용해 봅시다.

아래 메소드를 이용해서 nearest neighboer 모델을 만들고 **knn_model**에 저장하세요.

    graphlab.nearest_neighbors.create(학습데이터, features=사용하고자하는컬럼의리스트, label=각행의구분자역할을할컬럼명, distance=거리메트릭)
    
여기서는 피쳐로 'tfidf'만을 사용하고 label로는 이름을, distance는 'cosine'을 사용합니다.

In [41]:
knn_model = graphlab.nearest_neighbors.create(people, features=['tfidf'], label='name', distance='cosine')

Starting brute force nearest neighbors model training.

# 문서 조회에 nearest-neighbors 모델을 이용하기

## 누가 Obama에 가장 가까운가?

아래 메소드를 이용해서 Obama에 가까운 인물들을 찾아 보세요.

    model.query(SFrame)

In [42]:
knn_model.query(obama)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 24.285ms     |

| Done         |         | 100         | 676.317ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-------------------------+----------------+------+
| query_label |     reference_label     |    distance    | rank |
+-------------+-------------------------+----------------+------+
|      0      |       Barack Obama      |      0.0       |  1   |
|      0      |        Joe Biden        | 0.703138676734 |  2   |
|      0      |      Samantha Power     | 0.742981902328 |  3   |
|      0      |  Hillary Rodham Clinton | 0.758358397887 |  4   |
|      0      | Eric Stern (politician) | 0.770561227601 |  5   |
+-------------+-------------------------+----------------+------+
[5 rows x 4 columns]

Obama 대통령의 문서는 그의 부통령이었던 Biden과 가장 가깝고 그 다음으로는 Liberman과 같은 다른 정치인들이 등장하는 것을 확인할 수 있습니다.

## 다른 인물들로 문서 조회해 보기

'Taylor Swift'를 찾아서 **swift**에 저장해 봅시다.

In [43]:
swift = people[people['name'] == 'Taylor Swift']

Swift와 가까운 인물들을 찾아 봅시다.

    model.query(SFrame)

In [44]:
knn_model.query(swift)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 28.015ms     |

| Done         |         | 100         | 939.249ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+------------------+----------------+------+
| query_label | reference_label  |    distance    | rank |
+-------------+------------------+----------------+------+
|      0      |   Taylor Swift   |      0.0       |  1   |
|      0      |     Al Swift     | 0.699046316493 |  2   |
|      0      |    Bill Swift    | 0.712560304479 |  3   |
|      0      |    Ed Sheeran    | 0.735540453472 |  4   |
|      0      | Carrie Underwood | 0.736410347487 |  5   |
+-------------+------------------+----------------+------+
[5 rows x 4 columns]

이번에는 'Angelina Jolie'를 찾아서 **jolie**에 저장합시다. 

In [45]:
jolie = people[people['name'] == 'Angelina Jolie']

jolie와 가까운 인물들을 찾아 봅시다.

In [46]:
knn_model.query(jolie)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 19.84ms      |

| Done         |         | 100         | 792.555ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-----------------+----------------+------+
| query_label | reference_label |    distance    | rank |
+-------------+-----------------+----------------+------+
|      0      |  Angelina Jolie |      0.0       |  1   |
|      0      |    Brad Pitt    | 0.730168606132 |  2   |
|      0      | Barbara Hershey | 0.761752081266 |  3   |
|      0      |   Keith Jolie   | 0.770688506797 |  4   |
|      0      |  Julianne Moore | 0.776274899793 |  5   |
+-------------+-----------------+----------------+------+
[5 rows x 4 columns]

'Arnold Schwarzenegger'는 어떨까요? 먼저 해당 문서를 찾아서 **arnold**에 저장하세요.

In [47]:
arnold = people[people['name'] == 'Arnold Schwarzenegger']

이제 Arnold와 가까운 인물들을 찾아 봅시다.

In [48]:
knn_model.query(arnold)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 19.528ms     |

| Done         |         | 100         | 791.806ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+--------------------------+----------------+------+
| query_label |     reference_label      |    distance    | rank |
+-------------+--------------------------+----------------+------+
|      0      |  Arnold Schwarzenegger   |      0.0       |  1   |
|      0      |      Bonnie Garcia       | 0.79194972794  |  2   |
|      0      | Paul Grant (bodybuilder) | 0.803523487381 |  3   |
|      0      |       James Tramel       | 0.830161679751 |  4   |
|      0      |        Gray Davis        | 0.838121872529 |  5   |
+-------------+--------------------------+----------------+------+
[5 rows x 4 columns]

## 생각해 볼 거리

### 이 노트북에서는 왜 데이터를 split을 하지 않을까요?

# Quiz

## Quiz 1 
Elton John 문서에서 가장 많이 등장하는 단어들은 무엇인가요?

- 'Elton John'의 문서를 찾아서 저장합니다.
- 'word_count' 컬럼만으로 이루어진 SFrame을 만들고 stack 명령을 이용해서 'word'와 'count'로 분리하세요.
- 해당 결과값을 'count'로 내림차순으로 소팅하세요.

In [58]:
Jo = people[people['name'] == 'Elton John']

In [59]:
Jo.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: 1

Data:
+-------------------------------+------------+-------------------------------+
|              URI              |    name    |              text             |
+-------------------------------+------------+-------------------------------+
| <http://dbpedia.org/resour... | Elton John | sir elton hercules john cb... |
+-------------------------------+------------+-------------------------------+
+-------------------------------+-------------------------------+
|           word_count          |             tfidf             |
+-------------------------------+-------------------------------+
| {'all': 1, 'least': 1, 'pr... | {'all': 1.6431112434912472... |
+-------------------------------+-------------------------------+
[1 rows x 5 columns]

In [60]:
Jo[['word_count']].stack('word_count',new_column_name=['word','word_count']).sort('word_count',ascending=False)

Columns:
	word	str
	word_count	int

Rows: 255

Data:
+-------+------------+
|  word | word_count |
+-------+------------+
|  the  |     27     |
|   in  |     18     |
|  and  |     15     |
|   of  |     13     |
|   a   |     10     |
|  has  |     9      |
|  john |     7      |
|   he  |     7      |
|   on  |     6      |
| award |     5      |
+-------+------------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

## Quiz 2
Elton John 문서에서 가장 높은 TF-IDF 값을 가지는 단어들은 무엇인가요?

- 'tfidf' 컬럼만으로 이루어진 SFrame을 만들고 stack 명령을 이용해서 'word'와 'tfidf'로 분리하세요.
- 해당 결과값을 'tfidf'로 내림차순으로 소팅하세요.

In [61]:
Jo[['tfidf']].stack('tfidf',new_column_name=['word','tfidf']).sort('tfidf',ascending=False)

Columns:
	word	str
	tfidf	float

Rows: 255

Data:
+---------------+---------------+
|      word     |     tfidf     |
+---------------+---------------+
|    furnish    |  18.38947184  |
|     elton     |  17.48232027  |
|   billboard   | 17.3036809575 |
|      john     | 13.9393127924 |
|  songwriters  |  11.250406447 |
| tonightcandle | 10.9864953892 |
|  overallelton | 10.9864953892 |
|    19702000   | 10.2933482087 |
|   fivedecade  | 10.2933482087 |
|      aids     |  10.262846934 |
+---------------+---------------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

## Quiz 3

Eltion John과 Victoria Beckham의 문서 간에 TF-IDF로 표현된 거리는 어느 정도일까요?

In [63]:
VB = people[people['name'] == 'Victoria Beckham']

In [64]:
graphlab.distances.cosine(Jo['tfidf'][0],VB['tfidf'][0])

0.9567006376655429

## Quiz 4

Eltion John과 Paul McCartney의 문서 간에 TF-IDF로 표현된 거리는 어느 정도일까요?

In [66]:
PM = people[people['name'] == 'Paul McCartney']

In [68]:
graphlab.distances.cosine(Jo['tfidf'][0],PM['tfidf'][0])

0.8250310029221779

## Quiz 5

Eltion John과 Paul McCartney가 가까울까요? 아니면 Victoria Beckham이 가까울까요?

In [69]:
#### VB

## Quiz 6

word count를 이용해서 거리를 측정했을 때 Elton John에 가장 가까운 이웃은 누구일까요?
- word_count를 이용한 nearest_neighbors 모델을 만든다.
- query 메소드를 이용해서 최근접 이웃을 구한다.

참고: 위에서 만든 knn_model은 tf-idf를 이용해서 만들었기 때문에 word count를 이용해서 새로운 모델을 만들어야 합니다.

    graphlab.nearest_neighbors.create(학습데이터, features=사용하고자하는컬럼의리스트, label=각행의구분자역할을할컬럼명, distance=거리메트릭)
    
label은 인물의 이름을, distance는 'cosine'을 사용합니다.

In [70]:
wc_model = graphlab.nearest_neighbors.create(people, features=['word_count'], label='name', distance='cosine')

Starting brute force nearest neighbors model training.

In [71]:
wc_model.query(Jo)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 18.183ms     |

| Done         |         | 100         | 767.826ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-------------------+-------------------+------+
| query_label |  reference_label  |      distance     | rank |
+-------------+-------------------+-------------------+------+
|      0      |     Elton John    | 2.22044604925e-16 |  1   |
|      0      |   Cliff Richard   |   0.16142415259   |  2   |
|      0      |   Sandro Petrone  |   0.16822542751   |  3   |
|      0      |    Rod Stewart    |   0.168327165587  |  4   |
|      0      | Malachi O'Doherty |   0.177315545979  |  5   |
+-------------+-------------------+-------------------+------+
[5 rows x 4 columns]

## Quiz 7 

tf-idf를 이용해서 거리를 측정했을 때 Elton John과 가장 가까운 이웃은 누구인가요?
- 위에서 사용한 knn_model을 이용하면 됩니다.

In [72]:
knn_model.query(Jo)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 20.297ms     |

| Done         |         | 100         | 887.33ms     |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+------------------+--------------------+------+
| query_label | reference_label  |      distance      | rank |
+-------------+------------------+--------------------+------+
|      0      |    Elton John    | -2.22044604925e-16 |  1   |
|      0      |   Rod Stewart    |   0.717219667893   |  2   |
|      0      |  George Michael  |   0.747600998969   |  3   |
|      0      | Sting (musician) |   0.747671954431   |  4   |
|      0      |   Phil Collins   |   0.75119324879    |  5   |
+-------------+------------------+--------------------+------+
[5 rows x 4 columns]

## Quiz 8

word count를 이용해서 거리를 측정했을 때 Victoria Beckham에 가장 가까운 이웃은 누구일까요?

In [73]:
wc_model.query(VB)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 11.22ms      |

| Done         |         | 100         | 487.349ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+--------------------------+--------------------+------+
| query_label |     reference_label      |      distance      | rank |
+-------------+--------------------------+--------------------+------+
|      0      |     Victoria Beckham     | -2.22044604925e-16 |  1   |
|      0      | Mary Fitzgerald (artist) |   0.207307036115   |  2   |
|      0      |      Adrienne Corri      |   0.214509782788   |  3   |
|      0      |     Beverly Jane Fry     |   0.217466468741   |  4   |
|      0      |      Raman Mundair       |   0.217695474992   |  5   |
+-------------+--------------------------+--------------------+------+
[5 rows x 4 columns]

## Quiz 9 

tf-idf를 이용해서 거리를 측정했을 때 Victoria Beckham과 가장 가까운 이웃은 누구인가요?

In [74]:
knn_model.query(VB)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 16.519ms     |

| Done         |         | 100         | 699.736ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+---------------------+-------------------+------+
| query_label |   reference_label   |      distance     | rank |
+-------------+---------------------+-------------------+------+
|      0      |   Victoria Beckham  | 1.11022302463e-16 |  1   |
|      0      |    David Beckham    |   0.548169610263  |  2   |
|      0      | Stephen Dow Beckham |   0.784986706828  |  3   |
|      0      |        Mel B        |   0.809585523409  |  4   |
|      0      |    Caroline Rush    |   0.819826422919  |  5   |
+-------------+---------------------+-------------------+------+
[5 rows x 4 columns]